### DISCUSSION TITLE
Adding RTEs dynamically

### DISCUSSION THREAD

**Discussion by Fredericld** [livewire/livewire#2273] [1. Help]
Hello,

I am trying to build a simple page with questions and answers using livewire . I need my users to be able to add as many questions/answers as they like usig RTEs. Tinymce in my case

My html code
```
<div id="questions" class="tab-pane">
            <div class="row">
                <div class="col-lg-6">

                    @foreach($relatedQuestions as $key => $relatedQuestion)

                        <div class="form-group">
                            <input type="textarea" class="form-control tiny_question_title" placeholder="Enter title"  name="relatedQuestions[{{$key}}]['title']" wire:model.lazy="relatedQuestions.{{$key}}.title">
                            @error('relatedQuestions.'.$key.'.title')<span class="text-danger error">{{ $message }}</span>@enderror

                            <input type="textarea" class="form-control tiny_question_answer" placeholder="Enter text"  name="relatedQuestions[{{$key}}]['text']" wire:model.lazy="relatedQuestions.{{$key}}.text">
                            @error('relatedQuestions.'.$key.'.text')<span class="text-danger error">{{ $message }}</span>@enderror
                        </div>

                        <button class="btn btn-danger btn-sm" wire:click.prevent="removeRelatedQuestion({{$key}})">remove</button>

                    @endforeach

                    <button class="btn text-white btn-info btn-sm" wire:click.prevent="addRelatedQuestion({{$relatedQuestionsIteration}});">Add a question</button>

                </div>
            </div>
        </div>



@push('scripts')
<script>

    window.addEventListener('componentUpdated', event => {
        addTinyMCE();
    });

    // Add TinyMCE
    function addTinyMCE(){

        // Initialize
        tinymce.init({
            selector: '.tiny_question_title',
            themes: 'modern',
            height: 200,
            setup: function(editor) {
                editor.on('blur', function(e) {
                    @this.set('relatedQuestions.'.$relatedQuestionsIteration.'.title', tinymce.get('relatedQuestions.'.$relatedQuestionsIteration.'.title').getContent());
                });
            }
        });
    }

</script>
@endpush
```

My above code works to some extent. It adds a new Tinymce when clicking "add a  question", but I am not sure how to write the @this.set....  part
Does anyone know what to do so it saves the content in livewire?

**Comment by ju5t** [comment]
I would definitely suggest sponsoring @calebporzio, as he explains this concept in one of his screencasts + a lot more. 

In short, you could [combine @entangle with AlpineJS](https://laravel-livewire.com/docs/2.x/alpine-js) to link the two together. This is what I would suggest:

- Create a blade component for TinyMCE, so you can use something like `x-input.tinymce`.
- Enable AlpineJS on the component, you can do that with [x-data](https://github.com/alpinejs/alpine#x-data)
- You have extracted `addTinyMCE` already. You can use this in `x-data` as the example above, but I would change it, see below.

```javascript
function tinymce {
    return {
        init: function() {
            // Initialise TinyMCE here
            // ...
        }
    }
}
```

- Next, use [x-on](https://github.com/alpinejs/alpine#x-on) to initialise TinyMCE. You can use `x-on:(a js event)`. Link it to the `init()` function.

Now there will be a few changes to make in `init()`.

- The selector needs to be changed, instead of using a class name, use [x-ref](https://github.com/alpinejs/alpine#x-ref). To start, add `x-ref` to the text area, for example: `x-ref="tinymce"`. Keep in mind, AlpineJS is scoped, `x-ref` remains within the scope of that single component. That means you don't need to worry about it not being unique.
- Now, to link the two together so TinyMCE is initialised, in the selector itself you can use `$refs.tinymce`. Note: this will only work if you use AlpineJS.

At this point you should have the exact same thing as you have now and you should be able to use it like this:

```blade
<x-input.tinymce wire:model="relatedQuestions.{{ $key }}.title" />
```

This doesn't help you though; so now it's time to look at `@entangle`.

In your `x-data` function (`addTinyMCE` in your example), you need to add link the value of the input to the model in the Livewire component. You do this with:

```javascript
function tinymce {
    return {
        value: @entangle($attributes->wire('model'))
        ...
    }
}
```

I don't know for sure if you can use @entangle like that, but I think you can. Further more, this only works if you use a blade component, which is 100% recommended. 

This should link `value` property to the model. This is a two-way sync.

Now, in your TinyMCE on blur event:

`this.value = tinymce.get($refs.tinymce).getContent()`

I did most of this on the top of my head so I might be missing some things here and there. Neither do I know enough of TinyMCE to help you in depth with this, but it should get you going in the right direction.

**Reply by Fredericld** [reply]
@ju5t 

Sounds good. I will give this a try.
My code now almost works. I only experience issues with the delete functionality.
When I delete a question, I delete the correct element in the array of questions but the wrong RTE editor is removed. No matter which question I remove, it is always the last RTE that is removed. 
For each question I have a title and an answer. The answer is displayed in a textarea input field and updates properly. The title in tiny does not.

Does anyone know why tinymce is not refreshing properly??

```
    public function removeRelatedQuestion($id)
    {
        unset($this->relatedQuestions[$id]);
    }
```

  
```
                    @foreach($relatedQuestions as $key => $relatedQuestion)

                        <div class="form-group">
                            <div wire:ignore>

                                {!! Form::textarea('relatedQuestions['.$key.'][title]', '', array('placeholder' => 'Question','class' => 'form-control tiny_question_title', 'wire:model.lazy' => 'relatedQuestions.'.$key.'.title', 'id' => 'relatedQuestions['.$key.'][title]' )) !!}
                                @error('relatedQuestions.'.$key.'.title')<span class="text-danger error">{{ $message }}</span>@enderror
                            </div>
                            <input type="textarea" class="form-control tiny_question_answer" placeholder="Enter text"  name="relatedQuestions[{{$key}}]['text']" wire:model.lazy="relatedQuestions.{{$key}}.text">
                            @error('relatedQuestions.'.$key.'.text')<span class="text-danger error">{{ $message }}</span>@enderror
                        </div>

                        <button class="btn btn-danger btn-sm" wire:click.prevent="removeRelatedQuestion({{  $key }});">remove</button>

                    @endforeach

                    <button class="btn text-white btn-info btn-sm" wire:click.prevent="addRelatedQuestion();">Add a question</button>

```

**Reply by ju5t** [reply]
You're using HTML from LaravelCollective instead of [blade components](https://laravel.com/docs/8.x/blade#components). I can't help you with what you've made now, sorry.

As to removing things, based on what you pasted here I can only say that it should work.

**Reply by Fredericld** [reply]
@ju5t 

I tried to push my code as far as I could. 
Deleting is still an issue. The tinymce are not removed properly because of wire:ignore

I have tried to use alpine
```
                        <div id="questions" class="tab-pane @if ($activeTab == "questions") active @else fade @endif">
            <div class="row">
                <div class="col-lg-6">

                    @foreach($relatedQuestions as $key => $relatedQuestion)

                        <div class="form-group">
{{--                             
                            <div wire:ignore>
                                {!! Form::textarea('relatedQuestions['.$key.'][title]', '', array('placeholder' => 'Question','class' => 'form-control tiny_question_title', 'wire:model.lazy' => 'relatedQuestions.'.$key.'.title', 'id' => 'relatedQuestions['.$key.'][title]' )) !!}

                                @error('relatedQuestions.'.$key.'.title')<span class="text-danger error">{{ $message }}</span>@enderror
                            </div>
--}}

                <div wire:ignore>
                <textarea class="tiny" 
                        id='relatedQuestions[{{$key}}][title]'
                        name='relatedQuestions[{{$key}}][title]' 
                        rows="20"
                        wire:model = "relatedQuestions.{{$key}}.title"
                        wire:key="relatedQuestions.{{$key}}.title"
                        x-data
                        x-ref="relatedQuestions.{{$key}}.title"
                        x-init="
                        tinymce.init({
                                        selector: '.tiny',
                                        setup: function(editor) {
                                            editor.on('init change', function () {
                                                editor.save();
                                            });
                                            editor.on('blur', function(e) {
                                                myStr = tinymce.activeEditor.id;
                                                id = myStr.match(/\d+/);
                                                @this.set('relatedQuestions.'+id+'.title', tinymce.get('relatedQuestions['+id+'][title]').getContent());//tinymce.get('relatedQuestions.0.title').getContent());
                                            });
                                        }
                                    });
                        ">
                    </textarea>
                </div>

                            
                            <input type="textarea" class="form-control tiny_question_answer" placeholder="Enter text"  name="relatedQuestions[{{$key}}]['text']" wire:model.lazy="relatedQuestions.{{$key}}.text">
                            @error('relatedQuestions.'.$key.'.text')<span class="text-danger error">{{ $message }}</span>@enderror
                        </div>

                        <button class="btn btn-danger btn-sm" wire:click.prevent="removeRelatedQuestion({{ $key }});">remove</button>

                    @endforeach

                    <button class="btn text-white btn-info btn-sm" wire:click.prevent="addRelatedQuestion();">Add a question</button>

                </div>
            </div>
        </div>
```

and I get exactly the same result.  The text in the RTE does not get updated when one is deleted. I think now it all hinges on the @entangle line but I don't understand alpine enough to see where to put it. 
you mentioned I should use a component and I am going to learn that in the coming days.
